### DISCUSSION TITLE
Events for streams and connections when they close

### DISCUSSION THREAD

**Discussion by timprepscius** [microsoft/msquic#1611] [Q&A] [answered]
Hey there again,

This is probably in some doc, but I have not found it.

So let's say that I create a connection, and then immediately create a stream (without the connection first establishing itself).

Then let's say:

1.  The connection dies without establishing

Is either: QUIC_STREAM_EVENT_PEER_SEND_SHUTDOWN, or QUIC_STREAM_EVENT_PEER_SEND_ABORTED or QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE signaled?

Is QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_TRANSPORT signaled?

2.  Let's say the connection is established, but dies before the stream is created

Are any stream events signaled?  
I assume the connection shutdown events are signaled.

3.  Let's say the connection is established, and the stream is established.

I would assume both the stream events and the connection events are signaled.  True?

Thanks so much. 

This library is really clear so far.  I wish it did the simultaneous server+client on a socket, but since this doesn't actually solve all the edge cases, maybe it is better it doesn't work at all.

**Comment by nibanks** [comment, marked as answer]
I've been working on stream API documentation all day. Some of it is still in PR and some of it is merged. What you want is in the StreamStart.md and StreamShutdown.md files. Essentially, If a stream isn't successfully started, then it won't get any events delivered to it. It has to do with a race condition between calling StreamOpen and an event firing. Since that's hard/impossible to handle, events aren't delivered until the app calls StreamStart; but if you don't call it, you get no events. Please let me know if you need additional clarification.

**Reply by nibanks** [reply]
Oh, and if the connection is killed but the (started) stream itself was never explicitly shutdown, then you don't get any of the stream shutdown events, only the shutdown complete event. We can't give those events, because we don't have a stream shutdown error code to give you.

**Reply by timprepscius** [reply]
Thanks for this.  

I have a general programming question for you and perhaps process at MS.

Is there any "good" way of detecting race conditions besides just thinking a lot, talking, making lists of "this happens, then this happens, then this happens, but what if this is delayed?"

I often struggle with this, and usually the really bad race conditions- I only find through having some weird bug symptom which then leads to the unanticipated flow cause.

Is there an established process at MS for working out race conditions in advance, and how does it work?

**Reply by nibanks** [reply]
The only tried and true solution is testing. Write tests that explicitly exercise all known code paths. I know of no other real solution that will give you confidence in the code.

**Reply by timprepscius** [reply]
For this library do you write your tests before writing code, or after code, or both?

**Reply by nibanks** [reply]
At the same time. If a PR fixes a bug that was previously uncaught, it should include a test the validates the fix. Any new feature work also needs to include corresponding tests. Tests are extremely important.

**Reply by nibanks** [reply]
https://github.com/microsoft/msquic/blob/main/.github/CONTRIBUTING.md#tests

**Reply by timprepscius** [reply]
Cool, interesting.  Thanks again.
